addons/product/models/res_company.py

```diff
@@ -1,67 +1,47 @@
-                    'name': _("Default %(currency)s pricelist") %  params,
-                    'currency_id': new_company.currency_id.id,
-                })
-            self.env['ir.property']._set_default(
-                'property_product_pricelist',
-                'res.partner',
-                pricelist,
-                new_company,
-            )
+        companies = super().create(vals_list)
+        self._manage_companies_pricelists()
         return companies
 
-    def write(self, values):
-        # When we modify the currency of the company, we reflect the change on the list0 pricelist, if
-        # that pricelist is not used by another company. Otherwise, we create a new pricelist for the
-        # given currency.
-        ProductPricelist = self.env['product.pricelist']
-        currency_id = values.get('currency_id')
-        main_pricelist = self.env.ref('product.list0', False)
-        if currency_id and main_pricelist:
-            nb_companies = self.search_count([])
-            for company in self:
-                existing_pricelist = ProductPricelist.search(
-                    [('company_id', 'in', (False, company.id)),
-                     ('currency_id', 'in', (currency_id, company.currency_id.id))])
-                if existing_pricelist and any(currency_id == x.currency_id.id for x in existing_pricelist):
-                    continue
-                if currency_id == company.currency_id.id:
-                    continue
-                currency_match = main_pricelist.currency_id == company.currency_id
-                company_match = (main_pricelist.company_id == company or
-                                 (main_pricelist.company_id.id is False and nb_companies == 1))
-                if currency_match and company_match:
-                    main_pricelist.write({'currency_id': currency_id})
-                else:
-                    params = {'currency': self.env['res.currency'].browse(currency_id).name}
-                    pricelist = ProductPricelist.create({
-                        'name': _("Default %(currency)s pricelist") %  params,
-                        'currency_id': currency_id,
-                    })
-                    self.env['ir.property']._set_default(
-                        'property_product_pricelist',
-                        'res.partner',
-                        pricelist,
-                        company,
-                    )
-        return super(ResCompany, self).write(values)
+    def _manage_companies_pricelists(self):
+        """ Manage the default pricelists for all companies. """
+        if self.user_has_groups('product.group_product_pricelist'):
+            # Activate existing default pricelists
+            default_pricelists_sudo = self.env['product.pricelist'].sudo().with_context(
+                active_test=False).search([('item_ids', '=', False)]).filtered(
+                lambda pl: pl.company_id and pl.currency_id == pl.company_id.currency_id)
```

Comment:
I understand the correction, but now I'm wondering if:
```suggestion
            default_pricelists_sudo = self.env['product.pricelist'].sudo().with_context(
                active_test=False).search([('item_ids', '=', False)]).filtered(
                lambda pl: not pl.company_id or pl.currency_id == pl.company_id.currency_id)
```
wouldn't be better? So we'd keep any empty pricelist without company too. But then, where should we stop? Isn't a pricelist in another currency than the company one also a default pricelist that should be activated?

And what if we are activating a pricelist restricted to some countries only, but don't create one for all the others since we found a pricelist? Maybe restrict it to some country group too then... But then... what about the discount policy? and every other fields some modules added there? I don't particularly want to make a hook to get the domain searched here... :sweat_smile: